cmd/zscrub/zscrub.c

```diff
@@ -0,0 +1,260 @@
+/* Scrub a pool from userland */
+
+#include <stdio.h>
+#include <sys/zfs_context.h>
+#include <sys/spa.h>
+#include <sys/dmu.h>
+#include <sys/txg.h>
+#include <sys/dbuf.h>
+#include <sys/zap.h>
+#include <sys/dmu_objset.h>
+#include <sys/spa_impl.h>
+#include <sys/dsl_scan.h>
+#include <libnvpair.h>
+#include <libzfs.h>
+
+libzfs_handle_t *libzfs_handle;
+
+extern int reference_tracking_enable;
+extern int zfs_no_scrub_prefetch;
+extern enum ddt_class zfs_scrub_ddt_class_max;
+
+int verbose = 0;
+
+static void
+print_vdev(char *name, nvlist_t *nv, int depth)
+{
+	nvlist_t **child;
+	uint_t c, children;
+
+	vdev_stat_t *vs;
+	char *state;
+
+	if (nvlist_lookup_nvlist_array(nv, ZPOOL_CONFIG_CHILDREN,
+	    &child, &children) != 0)
+		children = 0;
+
+	VERIFY3U(nvlist_lookup_uint64_array(nv, ZPOOL_CONFIG_VDEV_STATS,
+	    (uint64_t **)&vs, &c), ==, 0);
+
+	state = zpool_state_to_name(vs->vs_state, vs->vs_aux);
+
+	(void) fprintf(stderr, "\t%*s%s %s\n", depth, "", name, state);
+
+	for (c = 0; c < children; c++) {
+		char *vname;
+		uint64_t islog = B_FALSE, ishole = B_FALSE;
+
+		/* Don't print logs or holes here */
+		(void) nvlist_lookup_uint64(child[c], ZPOOL_CONFIG_IS_LOG,
+		    &islog);
+		(void) nvlist_lookup_uint64(child[c], ZPOOL_CONFIG_IS_HOLE,
+		    &ishole);
+		if (islog || ishole)
+			continue;
+
+		vname = zpool_vdev_name(libzfs_handle, NULL, child[c],
+		    VDEV_NAME_TYPE_ID);
+		print_vdev(vname, child[c], depth + 2);
+		free(vname);
+	}
+}
+
+int
+main(int argc, char **argv)
+{
+	int error;
+
+	char **searchdirs = NULL;
+	int nsearchdirs = 0;
+	char *poolname = NULL;
+	int do_restart = 0;
+
+	spa_t *spa = NULL;
+
+	// Scan the entire DDT
+	zfs_scrub_ddt_class_max = DDT_CLASS_UNIQUE;
+
+	// Disable reference tracking debugging
+	reference_tracking_enable = B_FALSE;
+
+	// Disable prefetch during scan
+	zfs_no_scrub_prefetch = B_TRUE;
+
+	{
```

Comment:
Yeah, the extra space used by hard tabs makes the scoping braces a bit expensive. You're not *super* cramped here, so I don't mind if they stay.